kernel/src/hil/storage_interface.rs

```diff
@@ -0,0 +1,79 @@
+//! Interfaces for storage devices.
+
+use crate::returncode::ReturnCode;
+
+/// Cookies represent seekable positions within a storage interface. `SeekBeginning` allows a
+/// client to seek to the very beginning of the interface. `Cookie` allows the client to seek to a
+/// particular position within the interface. How a `Cookie` internally represents a position is up
+/// to the implementer.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum StorageCookie {
```

Comment:
It seems like storage cookie is both a public and private type. Since it has implementation-specific content, why not make it an associated type with the trait so there is no confusion that users shouldn't edit or create it?

Reply:
I'm not quite sure how that would work since the storage cookie type is used by both the `LogRead` and `LogWrite` traits. I'd also like to require that the storage cookie be an enum with the `SeekBeginning` variant, and I'm not sure how to do that if it's an associated type. Any recommendations on how to go about this? I agree that it would be much nicer if users weren't able to create storage cookies on their own.

Reply:
There is maybe a way to do that, since the flash hil has to do something similar? But, I don't think I have a good mental model for how the cookies should be used, and why the internal state of how to read the log needs to be exposed through the cookie type. If you could add a description of the intent of the cookie in the HIL that would be very helpful.

Reply:
I don't currently see a good way to refactor the `StorageCookie` enum as an associated type, so I'm going to leave it as is for now.